src/plcc/load_spec/validate_spec/validate_syntactic_spec/validate_ll1/generate_follow_sets.py

```diff
@@ -0,0 +1,70 @@
+from collections import defaultdict
+from .Grammar import Grammar
+from .generate_first_sets import FirstSetGenerator, generate_first_sets
+
+def generate_follow_sets(grammar: Grammar):
+    return FollowSetGenerator(grammar).generate()
+
+class FollowSetGenerator:
+    def __init__(self, grammar: Grammar):
+        self.grammar = grammar
+        self.firstSets = generate_first_sets(grammar)
+        self.followSets = defaultdict(set)
+        self.FirstSetGenerator = FirstSetGenerator(grammar)
+
+    def generate(self):
+        self.followSets[self.grammar.getStartSymbol()].add(self.grammar.getEOF().name)
+        updated = True
+        while updated:
+            updated = False
+            for nonterminal in self.grammar.getNonterminals():
+                if self._computeFollow(nonterminal):
+                    updated = True
+        return self.followSets
+
+    def _computeFollow(self, nonterminal):
+        updated = False
+        for lhs, rules in self.grammar.getRules().items():
+            for lists in rules:
+                for i, rule in enumerate(lists):
+                    if rule == nonterminal:
+                        if self._processFollowRule(lhs, lists, i, nonterminal):
+                            updated = True
+        return updated
+
+    def _processFollowRule(self, lhs, rules, index, nonterminal):
+        updated = False
+        if index + 1 < len(rules):
+            if self._addFirstOfNextSymbol(rules[index + 1], nonterminal):
+                updated = True
+        if index + 1 == len(rules) or self._canDeriveEmpty(rules[index + 1:]):
+            if self._addFollowOfLHS(lhs, nonterminal):
+                updated = True
+        return updated
+
+    def _addFirstOfNextSymbol(self, nextSymbol, nonterminal):
+        nextFirst = self.FirstSetGenerator._computeFirst(nextSymbol) - {self.grammar.getEpsilon().name}
```

Comment:
Here is the use of FirstSetGenerator. But we're violating its encapsulation to call _computerFirst(). Shouldn't firstSets have what we need?

I tried replacing this call with `self.firstSets[nextSymbol]` but it doesn't pass the tests. But I don't understand why.

I would really like to remove this strange dependency. Any ideas?

Reply:
@StoneyJackson I also ran into this same issue when refactoring these into separate files, and I believe that the tests do not pass strictly off the self.firstSets, because the firstSets only holds the FIRST sets for the nonterminals, but the FOLLOW set generation needs to utilize the FIRST sets for all the symbols, even terminals, which is done by using the _computeFirst() function. The way that this could be fixed is if the FirstSetGenerator's firstSet held the FIRST set for all symbols?